### ISSUE TITLE
`root_validator` methods are not called with subclasses as the first argument (always have the cls on which they were defined)

### ISSUE THREAD

**Issue by tlambert03** [pydantic/pydantic#5388]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2 installed directly from the `main` branch, or equivalent

### Description

I believe functions decorated `root_validator` are supposed to act like class methods, receiving the class as the first argument, and I would expect that to work with subclasses.  For example:

```python
In [3]: class A:
   ...:     @classmethod
   ...:     def method(cls):
   ...:         print(cls)
   ...:
   ...: class B(A):
   ...:     ...
   ...:

In [4]: A.method()
<class '__main__.A'>

In [5]: B.method()
<class '__main__.B'>
```

However, `root_validator` methods *always* appear to receive the class on which they were defined, and are not inherited... see example below.

### Example Code

```Python
from pydantic import BaseModel, root_validator

class Base(BaseModel):

    @root_validator(pre=True)
    def _some_validator(cls, values):
        print("cls", cls)
        return values

class Sub(Base):
    a: int = 1

Sub()
```

on pydantic v1, the above code snippet would print `'cls <class '__main__.Sub'>'`.  On v2 it prints `'cls <class '__main__.Base'>'`

### Python, Pydantic & OS Version

```Text
╰─❯ python -c "import pydantic.version; print(pydantic.version.version_info())"
             pydantic version: 2.0a1
                 install path: /Users/talley/dev/forks/pydantic/pydantic
               python version: 3.10.6 | packaged by conda-forge | (main, Aug 22 2022, 20:41:54) [Clang 13.0.1 ]
                     platform: macOS-11.7.4-x86_64-i386-64bit
     optional deps. installed: ['typing-extensions']
```

on `main` commit 0cd57b3e0c75d9622012de542e5ddc110fe55743
```

**Comment by dmontagu** [comment]
Great catch! Will let you know once this has been addressed.

**Comment by adriangb** [comment]
@tlambert03 should be fixed! The semantics did change a bit, so your input if you try it now would be appreciated.

**Comment by tlambert03** [comment]
Thank you!  Will have a look

**Comment by tlambert03** [comment]
works fine for my needs.  I didn't yet dig too deep into the implications for duplicate validator detection and stuff, but I definitely agree with the  "call whatever is under this attribute name on this class" semantics 👍 .  thanks for the fast turnaround!

**Comment by dmontagu** [comment]
@tlambert03 just fyi, the fix should be included now in the just-released `pydantic==2.0a2`. Please let us know if you notice any further issues.
